File: face/geomajas-face-dojo/client/src/main/resources/geomajas/spatial/geometry/util/RemoveRingOperation.js
/*
 * This is part of Geomajas, a GIS framework, http://www.geomajas.org/.
 *
 * Copyright 2008-2011 Geosparc nv, http://www.geosparc.com/, Belgium.
 *
 * The program is available in open source according to the GNU Affero
 * General Public License. All contributions in this program are covered
 * by the Geomajas Contributors License Agreement. For full licensing
 * details, see LICENSE.txt in the project root.
 */

dojo.provide("geomajas.spatial.geometry.util.RemoveRingOperation");
dojo.declare("RemoveRingOperation", GeometryOperation, {

	/**
	 * @fileoverview Operation for deleting a ring to a polygon or multipolygon.
	 * @class Extension of the {@link GeometryOperation}, for deleting a new ring
	 * to a {@link Polygon} or {@link MultiPolygon} object.
	 * @author Pieter De Graef
	 * 
	 * @constructor
	 * @extends GeometryOperation
	 * @param index Index to uniquely identify the polygon. Only needed if the
	 *              passed geometry is a {@link MultiPolygon}.
	 * @param ring The {@link LinearRing} object to add to the polygon.
	 */
	constructor : function (index) {
		this.index = index;
		
		/** Will store the deleted ring after execution of the {@link #edit} function */
		this.removed = null;
	},

	/**
	 * The actual deletion of the ring.
	 * @param geometry Should be of the type MULTIPLOYGON or POLYGON.
	 */
	edit : function (geometry) {
		if (geometry instanceof MultiPolygon) {
			if (this.index instanceof Array && this.index.length == 2) {
				var polygon = geometry.getGeometryN(this.index[0]);
				this.removed = polygon.getHoles().splice(this.index[1]-1, 1);
			} else if (this.index instanceof Array && this.index.length == 1) {
				var polygon = geometry.getGeometryN(0);
				this.removed = polygon.getHoles().splice(this.index[0]-1, 1);
			} else {
				var polygon = geometry.getGeometryN(0);
				this.removed = polygon.getHoles().splice(this.index-1, 1);
			}
		} else if (geometry instanceof Polygon) {
			if (this.index instanceof Array && this.index.length == 1) {
				this.removed = geometry.getHoles().splice(this.index[0]-1, 1);
			} else {
				this.removed = geometry.getHoles().splice(this.index-1, 1);
			}
		}
	},

	/**
	 * Can return the deleted ring after the {@link #edit} function has been
	 * called.
	 * @return A {@link LinearRing} object.
	 */
	getRemovedRing : function () {
		if (this.removed instanceof Array) {
			return this.removed[0];
		}
		return null;
	}
});

File: client/geomajas-client-smartgwt/client/src/main/java/org/geomajas/smartgwt/client/spatial/geometry/operation/RemoveRingOperation.java
/*
 * This is part of Geomajas, a GIS framework, http://www.geomajas.org/.
 *
 * Copyright 2008-2013 Geosparc nv, http://www.geosparc.com/, Belgium.
 *
 * The program is available in open source according to the GNU Affero
 * General Public License. All contributions in this program are covered
 * by the Geomajas Contributors License Agreement. For full licensing
 * details, see LICENSE.txt in the project root.
 */

package org.geomajas.smartgwt.client.spatial.geometry.operation;

import org.geomajas.smartgwt.client.spatial.geometry.Geometry;
import org.geomajas.smartgwt.client.spatial.geometry.LinearRing;
import org.geomajas.smartgwt.client.spatial.geometry.Polygon;

/**
 * <p>
 * <code>GeometryOperation</code> that removes an interior ring from a {@link Polygon} object.
 * </p>
 *
 * @author Pieter De Graef
 */
public class RemoveRingOperation implements GeometryOperation {

	/**
	 * The index of the ring to remove. Note that only interior rings can be removed! The exterior ring is never
	 * touched. If this index is smaller then 0, then the first interior ring is removed. If this index is too large,
	 * then the last interior ring is removed.
	 */
	private int ringIndex;

	/**
	 * Removes an interior ring from a {@link Polygon} object.
	 *
	 * @param ringIndex
	 *            The index of the ring to remove. Note that only interior rings can be removed! The exterior ring is
	 *            never touched. If this index is smaller then 0, then the first interior ring is removed. If this index
	 *            is too large, then the last interior ring is removed.
	 */
	public RemoveRingOperation(int ringIndex) {
		this.ringIndex = ringIndex;
	}

	/**
	 * Execute the operation! When the geometry is not a Polygon, null is returned. When the polygon does not have any
	 * interior rings, null is returned.
	 */
	public Geometry execute(Geometry geometry) {
		if (geometry instanceof Polygon) {
			Polygon polygon = (Polygon) geometry;

			// No rings? return null:
			if (polygon.getNumInteriorRing() == 0) {
				return null;
			}

			// Correct the index if necessary:
			if (ringIndex < 0) {
				ringIndex = 0;
			} else if (ringIndex >= polygon.getNumInteriorRing()) {
				ringIndex = polygon.getNumInteriorRing() - 1;
			}

			// Create the new array of interior rings:
			LinearRing[] interiorRings = new LinearRing[polygon.getNumInteriorRing() - 1];
			int count = 0;
			for (int n = 0; n < polygon.getNumInteriorRing(); n++) {
				if (n != ringIndex) {
					interiorRings[count++] = polygon.getInteriorRingN(n);
				}
			}
			return geometry.getGeometryFactory().createPolygon(polygon.getExteriorRing(), interiorRings);
		}
		return null;
	}

}
